pkg/controllers/management/node/controller.go

```diff
@@ -717,3 +742,81 @@ func (m *Lifecycle) updateRawConfigFromCredential(data map[string]interface{}, r
 	}
 	return nil
 }
+
+func (m *Lifecycle) deleteV1Node(node *v3.Node) (runtime.Object, error) {
+	logrus.Debugf("Deleting v1.node for [%v] node", node.Status.NodeName)
+	if nodehelper.IgnoreNode(node.Status.NodeName, node.Status.NodeLabels) {
+		logrus.Debugf("Skipping v1.node removal for [%v] node", node.Status.NodeName)
+		return node, nil
+	}
+
+	if node.Status.NodeName == "" {
+		return node, nil
+	}
+
+	// get user context k8s client to delete downstream cluster node
+	cluster, err := m.clusterLister.Get("", node.Namespace)
+	if err != nil {
+		return node, err
+	}
+
+	userClient, err := m.clusterManager.UserContext(cluster.Name)
+	if err != nil {
+		return node, err
+	}
+
+	err = userClient.K8sClient.CoreV1().Nodes().Delete(
+		context.TODO(), node.Status.NodeName, metav1.DeleteOptions{})
+	if !kerror.IsNotFound(err) {
+		return node, err
+	}
+
+	return node, nil
+}
+
+func (m *Lifecycle) drainNode(node *v3.Node) error {
+	cluster, err := m.clusterLister.Get("", node.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if !nodehelper.DrainBeforeDelete(node, cluster) {
+		return nil
+	}
+
+	logrus.Infof("node [%s] requires draining before delete, attempting to drain", node.Spec.RequestedHostname)
+	kubeConfig, err := m.getKubeConfig(cluster)
+	if err != nil {
+		return fmt.Errorf("node [%s] error getting kubeConfig", node.Spec.RequestedHostname)
+	}
+
+	if node.Spec.NodeDrainInput == nil {
+		logrus.Debugf("node [%s] has no NodeDrainInput, creating one with 60s timeout",
+			node.Spec.RequestedHostname)
+		node.Spec.NodeDrainInput = &v3.NodeDrainInput{
+			Force:           false,
+			DeleteLocalData: false,
+			GracePeriod:     60,
+			Timeout:         60,
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(m.ctx, time.Duration(node.Spec.NodeDrainInput.Timeout)*time.Second)
+	defer cancel()
+
+	_, msg, err := kubectl.Drain(ctx, kubeConfig, node.Status.NodeName, nodehelper.GetDrainFlags(node))
+	select {
+	case <-ctx.Done():
+		if ctx.Err() != nil {
+			logrus.Errorf("node [%s] drain failed: %s", node.Spec.RequestedHostname, ctx.Err())
+		}
+		if err != nil {
+			// kubectl failed continue on with delete any way
+			logrus.Errorf("node [%s] kubectl drain error: %s", node.Spec.RequestedHostname, err)
+		}
+	default:
```

Comment:
I don't think this empty default is necessary. If ctx is not done on select this will skip it, removing the effect of the timeout.

Reply:
great point